### ISSUE TITLE
关于数据处理和机器学习的一些问题

### ISSUE THREAD

**Issue by Truant-jason** [superhcb/DFYY#1]
1、在**train.py --97 line** 建议（对应其他模型也出现了这个问题）： 在数据处理的时候，先进行数据划分，再进行归一化. 因为fit_transform()方法会用到整个数据集的均值和标准差
     eg: 
                 
                 scaler = StandardScaler()
                 X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.3, random_state=42, stratify=y
    )
                 X_train = scaler.fit_transform(X_train)
                 X_test = scaler.transform(X_test)
2、除了smote之外，可以了解一下**加权交叉熵以及focal loss**，这里可以试一下加权交叉熵
3、样本不平衡，**可以保留f1比较高的模型的权重**，当前代码应该是训50轮 保留最后一轮
4、测试集指标计算的时候，要按照最好的YI对应的阈值点做cm，根据该阈值点计算对应的spe sen pre f1等指标。
5、pca做出来的特征可解释性比较差，他是将原始特征通过线性变换映射到一组彼此正交的新特征，**改变原先特征的意思，是特征降维技术**，可以使用matplotlib进行特征的可视化（比如说两维中，哪些特征聚成一堆，哪些聚成另一堆这样）。这里为了特征的可解释性，**可以考虑特征选择的技术**（特征降维和特征选择不一样），可以使用**Lasso**这个技术（需要调整α这个超参数（可以在训练集中用K折交叉验证进行调优），也可以自己手动调一下）
6、不是很建议使用knn这个算法，实验做下来感觉都比较偏低，不知道为什么，可能和算法的特性有关系。可以尝试其他的比如说基于梯度提升的一些算法，比如XGB / GBDT等等。 相应的一些参数配置可以单独写在一个**config.py或者yaml文件中**，方便修改
